lib/saml/post_url_service.rb

```diff
@@ -0,0 +1,89 @@
+# frozen_string_literal: true
+
+module SAML
+  # This class is responsible for providing the requests for the various
+  # SSO and SLO endpoints
+  # It provides a similar interface to URLService, but for most endpoints
+  # it returns an SSO URL and form request parameters for use in a SAML POST
+  # submission, instead of a self-contained redirect URL.
+  class PostURLService < URLService
+   # rubocop:disable Metrics/ParameterLists
+    def initialize(saml_settings, session: nil, user: nil, params: {},
+                   loa3_context: LOA::IDME_LOA3_VETS, previous_saml_uuid: nil)
+      unless %w[new saml_callback saml_logout_callback ssoe_slo_callback].include?(params[:action])
+        raise Common::Exceptions::RoutingError, params[:path]
+      end
+
+      if session.present?
+        @session = session
+        @user = user
+        @authn_context = user&.authn_context
+      end
+
+      @saml_settings = saml_settings
+      @loa3_context = loa3_context
+
+      if (params[:action] == 'saml_callback') && params[:RelayState].present?
+        @type = JSON.parse(params[:RelayState])['type']
+      end
+      @query_params = {}
+      @tracker = initialize_tracker(params, previous_saml_uuid: previous_saml_uuid)
+
+      Raven.extra_context(params: params)
+      Raven.user_context(session: session, user: user)
+    end
+    # rubocop:enable Metrics/ParameterLists
+
+    def should_uplevel?
+      user.loa[:current] < user.loa[:highest]
+    end
+
+    # rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+    def login_redirect_url(auth: 'success', code: nil)
+      if auth == 'success'
+        # if the original auth request specified a redirect, use that
+        redirect_target = @tracker&.payload_attr(:redirect)
+        return redirect_target if redirect_target.present?
+      end
+
+      # if the original auth request specified inbound ssoe and authentication
+      # failed, set 'force-needed' so the FE can silently fail authentication and NOT
+      # show the user an error page
```

Comment:
Could this be part of the method signature w/o major lift to refactor?

Otherwise, maybe this doc-comment can go above the method declaration, so it's visible in the generated YARD doc for the method.